ethexe/signer/src/digest.rs

```diff
@@ -0,0 +1,261 @@
+    }
+}
+
+impl fmt::Display for Digest {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "0x{}", hex::encode(self.0))
+    }
+}
+
+impl<T: ToDigest> FromIterator<T> for Digest {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        iter.into_iter().collect::<Vec<T>>().to_digest()
+    }
+}
+
+/// Trait for hashing types into a Digest using Keccak256.
+pub trait ToDigest {
+    fn to_digest(&self) -> Digest;
+}
+
+impl ToDigest for Digest {
+    fn to_digest(&self) -> Digest {
+        *self
+    }
+}
+
+impl<T: ToDigest> ToDigest for [T] {
+    fn to_digest(&self) -> Digest {
+        let mut message = Vec::with_capacity(self.len() * size_of::<Digest>());
+
+        for item in self.iter() {
+            message.extend_from_slice(item.to_digest().as_ref());
+        }
+
+        message.to_digest()
+    }
+}
+
+impl<T: ToDigest> ToDigest for Vec<T> {
+    fn to_digest(&self) -> Digest {
+        self.as_slice().to_digest()
+    }
+}
+
+impl ToDigest for [u8] {
+    fn to_digest(&self) -> Digest {
+        Digest(sha3::Keccak256::digest(self).into())
+    }
+}
+
+impl ToDigest for CodeCommitment {
+    fn to_digest(&self) -> Digest {
+        self.encode().to_digest()
+    }
+}
+
+impl ToDigest for StateTransition {
+    fn to_digest(&self) -> Digest {
+        // State transition basic fields.
+
+        let state_transition_size = // concat of fields:
+            // actorId
+            size_of::<Address>()
+            // newStateHash
+            + size_of::<H256>()
+            // valueToReceive
+            + size_of::<u128>()
+            // digest(valueClaimsBytes)
+            + size_of::<Digest>()
+            // digest(messagesHashesBytes)
+            + size_of::<Digest>();
+
+        let mut state_transition_bytes = Vec::with_capacity(state_transition_size);
+
+        state_transition_bytes.extend_from_slice(self.actor_id.to_address_lossy().as_bytes());
+        state_transition_bytes.extend_from_slice(self.new_state_hash.as_bytes());
+        state_transition_bytes.extend_from_slice(self.value_to_receive.to_be_bytes().as_slice());
+
+        // TODO (breathx): consider SeqHash for ValueClaim, so hashing of inner fields.
+        // Value claims field.
+
+        let value_claim_size = // concat of fields:
+            // messageId
+            size_of::<MessageId>()
+            // destination
+            + size_of::<Address>()
+            // value
+            + size_of::<u128>();
+
+        let mut value_claims_bytes = Vec::with_capacity(self.value_claims.len() * value_claim_size);
+
+        for ValueClaim {
+            message_id,
+            destination,
+            value,
+        } in &self.value_claims
+        {
+            value_claims_bytes.extend_from_slice(message_id.as_ref());
+            value_claims_bytes.extend_from_slice(destination.to_address_lossy().as_bytes());
+            // TODO (breathx): double check if we should use BIG endian.
```

Comment:
in solidity code it's definitely big endian
```suggestion
```

Reply:
Done